OpenRC service scripts

rseichter opened the pull request:
Having recently inherited the maintainership for Gentoo Linux's OpenDKIM package, I suggest adding our [OpenRC](https://github.com/OpenRC/openrc) service scripts.

orlitzky commented:
Thanks Ralph, I was going to put together this list sometime later today but you were too fast.

There's a lot we need to clean up before we adding these... I'll try to list them in a sensible order.

**1. Variable replacement**

The magic variables like `@sbindir@` are replaced at build-time by `./configure`, but not unless you tell it to. The usual way is to rename the "source" file with an `.in` suffix, like `opendkim.initd.in`, and then that file needs to be added to the bottom of `configure.ac` so that autoconf knows to process it (it will replace the variables, and save the output without the `.in` suffix).


**2. Use of runstatedir**

Then, we should depend on a newer version of autoconf. At the top of configure.ac, there's a line like

```
AC_PREREQ(2.61)
```

which says what version needs to used, at a minimum. (This is only for developers). In autoconf-2.70, a new variable "runstatedir" was added:

  https://www.gnu.org/prep/standards/html_node/Directory-Variables.html

and that's what we want to use instead of `/run` everywhere, so I think we should depend on autoconf-2.70.


**3. Add a tmpfiles.d entry**

The `@runstatedir@/opendkim` directory needs special permissions, and a tmpfiles.d entry can do that for both OpenRC and systemd:

```
d @runstatedir@/opendkim 0755 opendkim opendkim
```

**4. Using the new runstatedir to store the pid file and socket.**

In the init script (and in upstream's system service file -- but we can worry about that later!), there are some things we can improve now that `@runstatedir@` is available. The spirit of these is the same as the suggestions in my comment 5  on https://bugs.gentoo.org/606978, but instead of just declaring `/run/opendkim` to be the socket path by fiat, here we're constructing it from `@runstatedir@` at build time.

```sh
pidfile="@runstatedir@/${RC_SVCNAME}.pid"
```

(Note: `RC_SVCNAME` is still fine, no need for `$prog`)

And we can get rid of _all_ of the socket handling stuff:

```sh
command_args="-P ${pidfile} -x ${CONFFILE}"
...
start_pre() {
	# If this isn't a restart, make sure that the user's config isn't
	# busted before we try to start the daemon (this will produce
	# better error messages than if we just try to start it blindly).
	#
	# If, on the other hand, this *is* a restart, then the stop_pre
	# action will have ensured that the config is usable and we don't
	# need to do that again.
	if [ "${RC_CMD}" != "restart" ]; then
		checkconfig || return $?
	fi
}
```

We can get rid of the conf.d file now too; it's not needed if OpenRC/systemd agree to use `@runstatedir@/opendkim` for local sockets, and if the configuration file is the one source of truth for the socket path.

The stale socket handling has also been removed above, since we don't know the socket path. This is fine: the init system shouldn't be doing that anyway, and systemd doesn't do it. Stale sockets only get created in exceptional circumstances anyway, and when it happens, maybe you don't want your init system just deleting shit to hide the problem.


**5. Configuration file updates**

Now that "runstatedir" is gospel, we should put an example local socket entry in the config file, telling people that their local sockets should go in `@runstatedir@/opendkim`. The example configs are already processed by autoconf, so in e.g. `opendkim.conf.simple.in` we could add

```ini
# To use a local socket instead, specify a filename under @runstatedir@/opendkim.
# Socket  local:@runstatedir@/opendkim/opendkim.sock
```

**6. Install process**

Rather than document the files that need to be copied and where in a `README`, we should have e.g. `make install-openrc` do it.

**7. Conclusion / summary of goals**

Ultimately, both systemd and OpenRC will have `@runstatedir@/opendkim` created automatically, with the right permissions, by the tmpfiles.d entry. The socket would be set in the config file (where the correct path based on `@runstatedir@` is shown), and then regardless of whether it was inet/local and regardless of the init system, things would just work.

Basically, we should be able to `make && make install && make install-openrc` on Gentoo after setting the correct `--runstatedir=/run` at build time, and have the resulting init script work out of the box, and have the systemd service also work out of the box with no changes.

It does seem like a lot of work, but supporting systemd forces us to do a few things simpler, and even though I'm a systemd hater, I think the simplicity is a good thing. With systemd's tmpfiles entry in place, we can just re-use it in OpenRC, greatly simplifying things there, too. Then it's just a matter of undoing some old cruft to bring the two to parity. And then upstreaming everything to get the cool `@vars@`.

rseichter commented:
@orlitzky Please see my email.

orlitzky committed:
contrib/systemd: add tmpfiles.d configuration file.

The tmpfiles.d mechanism, described at

  https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html

is a way to tell systemd (and these days, OpenRC) what directories
need to be created at boot time, and with what permissions/ownership.
The existing OpenDKIM systemd service file (in contrib/systemd), for
example has the following lines,

  PIDFile=@localstatedir@/run/opendkim/opendkim.pid
  User=opendkim
  Group=opendkim

Since the daemon starts as user "opendkim" it will not, in general, be
able to write to the directory specified in the PIDFile directive. This
commit introduces a tmpfiles entry to fix that: the @runstatedir@/opendkim
directory will be created at boot-time, and owned by the "opendkim" user.
(The systemd service file above requires some more work, independent of
this change, that we can do at a later time.)

This change also benefits OpenRC, whose "opentmpfiles" implementation can
use the same tmpfiles.d entries as systemd. On operating systems where
both init systems are used, having the same tmpfiles entry will allow the
init systems to share one directory for, say, a local socket. An OpenRC
service file is planned for a later commit.

Gentoo-bug: https://bugs.gentoo.org/606978

orlitzky committed:
contrib/systemd: use @runstatedir@ in opendkim.service.in.

Now that the "runstatedir" variable is supported by the build system,
use it in the systemd service file template. The existing PID file
path was under "$@localstatedir@/run", and this commit changes that
string to "@runstatedir@". Unless the user goes out of his way to
specify a custom runstatedir, these should be equivalent: the value of
runstatedir in configure.ac defaults to $localstatedir/run.

rseichter committed:
configure.ac: Added workaround to support @runstatedir@

On modern systems /run replaces /var/run, so using @runstatedir@
instead of @localstatedir@/run is preferable.

Autoconf 2.70 will support @runstatedir@ , but it has not been
released yet, so a workaround is required for version 2.61.

People can start using @runstatedir@ and once autoconf 2.70 becomes
available, the workaround can be dropped and AC_PREREQ updated.

rseichter committed:
Added example for local socket using @runstatedir@

rseichter committed:
Added Gentoo Linux's OpenRC service scripts.

See https://github.com/OpenRC/openrc for more information.

orlitzky committed:
contrib/systemd: enable full @VARIABLE@ substitution.

The existing @variable@ substitutions in contrib/systemd don't quite
work, and the autoconf documentation recommends against the current
approach for the same reason: unless you specify a full, absolute,
path for (say) @sbindir@, then autoconf will only expand the default
value of that variable one level, to "${prefix}/sbin". That's not
usable out of the box, and we have no good way to replace the
"${prefix}" variable with its configure-time value.

This commit switches the substitution tactic to perform the
replacement in the Makefile, rather than with autoconf. This is in
fact what the autoconf documentation recommends, even if it looks
silly. Instead of (say) @sbindir@, we use a custom @SBINDIR@ variable,
and then set it to the appropriate value with sed at build-time. The
value of the variable itself still comes from the build system.

orlitzky committed:
opendkim/opendkim.conf.*.in: use @runstatedir@ in PidFile paths.

Now that the @runstatedir@ variable is available, use it in the
(commented-out) PidFile settings present in the two sample
configuration files.

The example location for these two files has also been moved out of
the "opendkim" subdirectory of @runstatedir@ for security reasons.
Generally, that subdirectory will be writable by a restricted user as
whom the daemon is running, so that, for example, a local socket can
be created there. The PID file, on the other hand, needs to be
writable only by root, because root will generally be the one who
calls "kill" on its contents. OpenDKIM handles this situation
correctly: the PID file is created as root, even when dropping
privileges, so the PID file is successfully written directly to
@runstatedir@.

orlitzky committed:
opendkim/opendkim.conf.*.in: enable full @VARIABLE@ substitution.

The existing @runstatedir@ substitutions in the example configuration
files don't quite work, and the autoconf documentation recommends
against the current approach for the same reason: unless you specify a
full, absolute path for @runstatedir@, then autoconf will only expand
the default value of that variable one level, to "${localstatedir}/run".
That's not usable out of the box, and we have no good way to replace
the "${localstatedir}" variable with its configure-time value.

This commit switches the substitution tactic to perform the
replacement in the Makefile, rather than with autoconf. This is in
fact what the autoconf documentation recommends, even if it looks
silly. Instead of @runstatedir@, we use a custom @RUNSTATEDIR@
variable, and then set it to the appropriate value with sed at
build-time. The value of the variable itself still comes from the
build system.

orlitzky committed:
contrib/systemd: fix inconsistencies in the systemd service file.

A bunch of not entirely related improvements to the systemd service:

 * Line-wrap the comment at the top, and provide better instructions
   for what to do in case you need to depend on a database. End users
   should be appending configuration, and not editing it in-place.

 * Line-wrap the "Documentation" directive.

 * Remove the nonexistent opendkim-testadsp(8) manpage from the list
   of Documentation.

 * Move the PIDFile out of the "opendkim" subdirectory of @RUNSTATEDIR@
   for security reasons. The contents of the PID file will be killed as
   root, so it's important that it live in a root-owned directory so as
   not to be modified by a non-root user.

 * Force the PID file to be where we'll look for it, with the "-P"
   flag on the command line.

 * Drop the User/Group directives. With Type=Forking, the OpenDKIM
   daemon will drop privileges itself, to the user and group specified
   in its own configuration file.

 * Add an ExecStartPre check to ensure that the configuration file is
   valid before attempting a restart (that might otherwise fail).

 * Drop the "EnvironmentFile" directive and the "$OPTIONS" variable.
   These are not portable; the systemd folks regret adding them and
   tell you not to use them. If additional configuration is needed,
   the official way to do it is by appending service configuration
   files.

orlitzky committed:
autotools: refactor build-time @VARIABLE@ substitutions.

Some recent commits introduced build-time variable substitution in
three places,

  1. contrib/OpenRC/*
  2. contrib/systemd/*
  3. opendkim/opendkim.conf.*.in

Each of these uses essentially the same strategy, making it a good
candidate to be factored out. This commit does just that, moving
everything into the top-level Makefile.am, which is the closest parent
of both the "contrib" and "opendkim" directories.

Ultimately, the behavior of the rules is unchanged.

orlitzky commented:
@mskucherawy 

I believe this is ready for serious consideration now. Most of the points I brought up in my earlier comment have been addressed:

1. We've added backwards-compatible `runstatedir` support to the build system in anticipation of the next version of autotools.
2. We've added an example to the simple configuration file, showing people the best place to put their local socket (i.e. one that init systems can anticipate and prepare for).
3. We've added an OpenRC service script. Gentoo is the primary source of OpenRC users, and Ralph and I are responsible for OpenDKIM on Gentoo and are pretty confident that this script is in good shape. (In other words, I'm asking you to just trust us that it works if you don't run OpenRC yourself.) I promise to fix any bugs!
4. To create the socket directory automatically at boot time in OpenRC/systemd, we've added a tmpfiles.d entry...
5. ...and while we were at it, some improvements were made to the systemd service file. Hopefully these are uncontroversial; they are also tested on Gentoo, and happen to fix your issue #38.
6. For the files this PR touches, we've implemented a better way to do full `@variable@` substitution. Using e.g. `AC_SUBST` in `configure.ac` doesn't always substitute the full path, and a better way is suggested by the autotools documentation (in the "Installation Directory Variables" section). We haven't gone through and updated all of the other files that were using autoconf-based substitutions, but they are easily fixable now that the right framework is in place.

A few things we **didn't** fix:

* The `opendkim` user was hard-coded for the systemd stuff before, and I've left that alone. We hard-code it for OpenRC now, too. Some time in the future it can be made configurable. The new variable substitution framework should make it easy.
* Everything is still installed as documentation. To do it right, we would want people to pass e.g. `--enable-openrc` or `--enable-systemd` to the configure script before we started installing those scripts. OpenRC in particular installs to `/etc/init.d` which might belong to some other piece of software. In any case, they're  still installed -- just as documentation for now. As a packager this isn't a big deal, we just move it to the right place afterwards.

Our Gentoo users are running the patch corresponding to this PR, and so far no one has complained. Ralph and I have also done a lot of testing. Carrying the patch isn't a big deal, but it makes it hard to pile on future improvements, so I hope we can get some feedback soon. Thanks for your time!

rseichter commented:
@mskucherawy This pull request has now been open for more than a year. Some form of reaction would be appreciated.

mskucherawy commented:
LGTM; are we going to add the RELEASE_NOTES text in here or in a later pull request?

rseichter commented:
@mskucherawy I'm not certain if this is a question you are asking your co-reviewers, or @orlitzky and myself? Michael has provided extensive information in a comment above, and I doubt there is much room for improvement left. Could you use Michael's comment as a basis for the release notes?

orlitzky commented:
Most of these changes are invisible to end users. How in-depth would you like the release notes to be? Something like,

* Fixed `@variable@` substitution in the systemd service file.
* Added a new tmpfiles.d entry to create the directory where a local socket would live.
* Moved the PID file to the "runstatedir", specified with e.g. `make runstatedir=/run` for now. The default is backwards-compatible.
* Added an OpenRC service script that is installed unconditionally as documentation.

In other news, autoconf-2.70 is (finally) almost ready. After it's widely available, we can require it in `configure.ac` to improve the `@runstatedir@` user experience.

rseichter commented:
@orlitzky Do you want to do anything with the data in this pull request? I have lost all interest after the many years which have passed.

orlitzky commented:
No, you can close it if you're tired of looking at it. We still have the patch in Gentoo if someone wants to track it down.

rseichter commented:
Closing PR due to lack of activity.
